### PR TITLE
Treat store nodes specially in CSE to avoid breaking +=

### PR DESCRIPTION
**Pull request by abadams**
@aekul and I have both found, on different projects using lots of random schedules, that large summations (or other in-place reductions) into Funcs that are compute_root have unreliable performance in the sense that small changes to the compiler or to the schedule make them slower or faster for no apparent reason. Sometimes the summation is done into a register and stored once at the end, and sometimes there are loads and stores to memory in the loop. The culprit seems to be instances where f[x] = f[x] + y gets CSE'd into f[x] = let ... let ... in f[z] + y, in cases where x and y share a common subexpression. After LICM it's no longer apparent to llvm that x and z are in fact equal, especially if they both just got hoisted out of a GPU kernel as separate args.

This PR does CSE jointly on the index and the value in a store node instead, and it seems to fixes the cases that I have encountered on CPU and the cases that @aekul has encountered on GPU so far.

This mostly doesn't affect good human-written schedules, as humans tend to do the summation into something explicitly stored in registers.

Let's see what the buildbots think of this change.

### CODE REVIEW

File: src/CodeGen_LLVM.cpp [hunk 1]
```diff
@@ -1804,6 +1804,17 @@ Value *CodeGen_LLVM::codegen_buffer_pointer(Value *base_address, Halide::Type ty
         index = promote_64(index);
     }
 
+    // Peel off a constant offset as a second GEP. This helps LLVM's
+    // aliasing analysis, especially for backends that do address
+    // computation in 32 bits but use 64-bit pointers.
```
Comment (steven-johnson) [inline src/CodeGen_LLVM.cpp:1809]: out of curiosity, what backends are those? Seems like an odd choice.

Comment (abadams) [reply src/CodeGen_LLVM.cpp:1809]: CUDA. pointers are 64-bit but 32-bit integer arithmetic is cheaper than 64-bit

File: src/CSE.cpp [hunk 2]
```diff
@@ -228,6 +228,31 @@ class RemoveLets : public IRGraphMutator {
...
+    using IRMutator::visit;
+
+    Stmt visit(const Store *op) override {
+        // It's important to do CSE jointly on the index and value in
+        // a store to stop:
+        // f[x] = f[x] + y
+        // from turning into
+        // f[x] = f[z] + y
+        // due to the two equal x's indices being CSE'd differently due to the presence of y.
+        Expr dummy = Call::make(Int(32), std::string{}, {op->value, op->index}, Call::PureIntrinsic);
```
[the remark this answers is not in the document]
Comment (abadams) [reply src/CSE.cpp:240]: Maybe I should add a pure intrinsic that just means these exprs are bundled together for the purpose of analysis?

[the remark this answers is not in the document]
Comment (steven-johnson) [reply src/CSE.cpp:240]: SGTM, then you could add a check in codegen_llvm (etc) to assert that you never see it

[the remark this answers is not in the document]
Comment (abadams) [reply src/CSE.cpp:240]: Done. CodeGen_LLVM already errors on unhandled intrinsics.

**Comment by steven-johnson** [comment]
LGTM (again)

**Pull request merged**